STYLE.md

```diff
@@ -0,0 +1,262 @@
+# Vector Style Guide
+
+As a large, open-source project, it can be a struggle to ensure a consistent level of code quality.  This style guide is
+meant to be the canonical reference for all things style: code comments, formatting, acceptable (or unacceptable)
+crates, data structures, or algorithms, and so on.
+
+In essence, we hope to turn pull request review comments like "why did you do it this way?" or "I think you could try
+doing it this way" into "we always do X this way: <link to style guide>".
+
+# Formatting
+
+At a high-level, code formatting is straightforward: we use the native `rustfmt` exclusively.
+
+You can acquire `rustfmt` -- which is invoked as `cargo fmt` -- by following the directions listed out [in the rustfmt
+repository](https://github.com/rust-lang/rustfmt#on-the-stable-toolchain).
+
+Vector has its own formatting rules (`.rustfmt.toml`) that will automatically be used when you run `cargo fmt` within
+the repository.  If you're ever in doubt, you can also run `make check-fmt` which will invoke `cargo fmt` and ensure all
+code follows our formatting rules.
```

Comment:
`rustfmt` doesn't always format code within macros, in which case you should make a best effort to make the code formatted like the code in the rest of the project.

Reply:
That's a good point. I think we could maybe add a note to be on the lookout, although I don't want people to think they always _have to be_ on the lookout. Ideally it'd be something we could lightly suggest in reviews if we spot it.

Either way, I'll try to add a small note or reword it to indicate that `cargo fmt` is the bare _minimum_ but there might need to be other manual tweaks done to fit within the expected formatting style.

Reply:
Yeah, it's just something that surprised me a couple of times when I started doing Rust. I don't we need any more than a small note to point out that it's because the code is in a macro that the formatting is not working as expected.

Reply:
Addressed in ae129d597.